src/core/lib/json/json_reader.cc

```diff
@@ -157,9 +157,10 @@ bool JsonReader::StringAddChar(uint32_t c) {
 }
 
 bool JsonReader::StringAddUtf32(uint32_t c) {
-  if (c <= 0x7f) {
-    return StringAddChar(c);
-  } else if (c <= 0x7ff) {
+  // (b/208085307): This argument of this function is an Utf32 encoded
+  // character. So we need to treat it as such and not as a simple ASCII
+  // character even if it is <= 0x7f.
+  if (c <= 0x7ff) {
```

Comment:
I'm not sure this is right... wouldn't this change imply that it's impossible to encode a 1-byte long utf-8 symbol with utf-32?

Reply:
Yes I think you are right. But there might be another issue.

The json writer simply outputs all chars >= 32 and <= 127 without any modifications: https://github.com/grpc/grpc/blob/5961aeb2a02c280a67efa9247b2b77304af865c8/src/core/lib/json/json_writer.cc#L145

But for chars < 32 or chars >= 127, it applies utf-16 encoding and creates substrings of the form "\u<encoded bytes>".  From my understanding of the json reader, the function in question in the PR (i.e StringAddUtf32), is called only when parsing substrings of the form "\u<encoded bytes>" .

I think there  is some ambiguity during parsing of strings of the form "\u<encoded_bytes>". 
Sometimes the "<encoded_bytes>" are less <= 127 while the original char is < 32 while other times the "<encoded_bytes>" are <= 127 while the original char is >= 127. 
So when <encoded_bytes> is <= 127, it is not clear how the original char should be interpreted.

I think this can be corrected by not doing Utf16 encoding here (for chars < 32 or char == 127) and coupling it with the current change already in the PR: https://github.com/grpc/grpc/blob/5961aeb2a02c280a67efa9247b2b77304af865c8/src/core/lib/json/json_writer.cc#L166

WDYT ?